pgxn/neon/walproposer.c

```diff
@@ -1104,6 +1053,9 @@ SendProposerElected(Safekeeper *sk)
 	term_t		lastCommonTerm;
 	int			i;
 
+	/* Now that we are ready to send it's a good moment to create WAL reader */
+	wp->api.wal_reader_allocate(sk);
```

Comment:
nit: it's strange that it is deallocated inside `conn_finish()`. I'd try to think how to make it clearer, or at least add comment to the `conn_finish()` documentation, that it should deallocate wal_reader.

Reply:
It's messy, yes, and again about drawing border between walproposer and walproposer_pg. The idea is: due to potentially existing remote connection xlogreader is not monotonic, you can't request reads at random LSNs. Since ShutdownConnection can be reached though many trajectories, e.g. not neon_walreader failure but sk error, it is simpler not to try to figure out whether xlogreader can be reused and simply reset it. And since performance of reset vs reallocate here doesn't matter, there is no special reset at all, just reallocate.

In principle we could add yet another API call for xlogreader deallocation. But it requires changes in many places (api, walproposer_pg implementation, api_bindings.rs, rust API trait, rust API trait implementation) and makes code harder to read/follow. Do we really need that if that's called exactly at conn_finish? Seems like no...

Reply:
Maybe it's possible to make NeonWALReader handle non monotonic reads (e.g. reset state on such), then we don't need to deallocate it. But for now comment is enough.